### DISCUSSION TITLE
Cross Platform Struct Interface with Fixed Size for High performance Value Types

### DISCUSSION THREAD

**Discussion by ackava** [dotnet/csharplang#3175] [Language Ideas]
**Update**

> After reviewing many answers and learning about internals, I found out that it is not possible, as reference types inside `struct` needs to be garbage collected and it is only possible if JIT knows the type instead of interface. Somehow, I was looking for light weight classes that can live only on the stack, this is possible with `ref struct` but applications are still limited and there is no real benefit of struct interface. And generics can handle structs correctly.
> 
> As we are moving towards .Net Standard, where we do not use platform dependent code and rely on dependency injection, it causes serious performance issue for value type (struct) implementations.
> 
> This discussion can be closed.
> 

Currently all value types are passed as an object when referenced through an interface (boxing). Mostly in platform independent code, we use interface to create a class which holds single field and all operations are simply forwarded to native implementation.

## Special Attribute `[StructSize(n)]`

Simple extra attribute on interface, `[StructSize(int)]`.

```c#

   [StructSize(4)]
   public interface IColor {
   }

   public struct NSColor: IColor {
   }

   public struct DroidColor: IColor {
   }

```

All interfaces declared with `[StructSize(4)]` can only be implemented by `struct` and not `class`. All code that accesses `IColor` will not be boxed. These interfaces will be sealed as further any other interface cannot inherit this one. And one class can only implement one such interface.

## OR (Alternative) Special Interfaces

So lets assume there are these types,
`IValueBytes8, IValueBytes16, IValueBytes32` types in the system where else they mention how much bytes they occupy. In each type, first 4 bytes specify the type pointer used by `GetType()`, all methods can never be virtual and they will not be part of data bytes. So JIT knows that it needs 8+4 = 12 bytes for IValueBytes8.

Structs implementing IValueBytesN should only contain bytes less than or equal to the specified size.  Actual Type will than have implemented method pointers for invoking implementation method.

```c#
     public interface IColor: IValueBytes8 {
     }

     // windows implementation
     // this can have maximum of 8 bytes of information
     public struct WinColor: IColor {
        // ...
     }
    
     // not allowed !!!!! this should give error
     public class WinClass: IColor {
     }
```

```c#
public class SomeConsumerService {

    public IColor Convert(IColor input) {
        // these interfaces will not be of type object, they will be value types
        //   and will be treated as value type to avoid boxing...
    }

    public Nullable<IColor> Convert(Nullable<IColor> input) {
          // they can be null only if they are declared as nullable...
          // this certainly will be boxed
    }

}
```

# Benefit are

1. We can easily write high performance library having simple interface whose implementations can be value types in different platforms. This will reduce GC overhead.
2. We can write unit tests.
3. JIT can inline code to improve performance further.
4. Code can be free of Generics
5. Some extra reflection method can also help in handling such types while calling reflection methods.
6. You also may avoid reflection and deal with interfaces easily.

### Problem with Generic Constraint

Problem with generic constraint is every class method that accepts/returns must add one more generic parameter.

Lets assume, we have three projects, 
1. Library.Core (.net standard 2.1), some interfaces with generics
2. Library.iOS (Xamarin.iOS) contains implementation of Library.Core
3. Library.Droid (Xamarin.Droid) contains implementation of Library.Core

4. CommonConsumer (.net standard) references only Library.Core

Problem is, `CommonConsumer` have to write too much of generics in order to maintain struct being converted to object. This becomes even more complicated that everything has to be generic even though not required.

Now lets look at sample code, 

```c#
public class SomeConsumerService<T>
    where T: struct, IValueType 
{

    public T Process(T input) {
        // these interfaces will not be of type object, they will be value types
        //   and will be treated as value type to avoid boxing...
    }

    public Nullable<T> Process(Nullable<T> input) {
          // they can be null only if they are declared as nullable...
    }

}
```

See the amount of code changed from simple class to one with Generics, Add complication of dynamically invoking, inspecting parameter and injecting values etc. When you add multiple hierarchies it just increases code complexity.

I don't know which implementation will be easy, but CLR and Roslyn team can identify best suitable and less expensive proposal.

**Comment by DaZombieKiller** [comment]
Reader expectations of code aside (boxing, etc) this seems like it largely defeats the purpose of an interface, and I'm not seeing a huge benefit over the constraint method for your given example (all three of those listed benefits are possible today, too):

```CSharp
public T? Process<T>(T? input)
    where T : struct, ISomeStruct
{
}
```

Could you provide an example that showcases this being 'too much code'? I assume you're talking about a situation where you have a method taking multiple different interface types.

This is likely a situation that would be far better served by the shapes proposal.
EDIT: You could also accomplish this using an attribute and a custom analyzer to provide warnings.

**Comment by ackava** [comment]
I understand it is possible, but imagine a library as huge as `Xamarin.Forms`, every method, every class I need to write will have unnecessary extra `<T>`.

I already have a library and have nearly 100 methods with 20 classes, and allowing other developers to extend this classes and unnecessary write extra `<T>`.

Problem occurs when I have to declare instance methods on class as specified here. As this class is injected as dependency, I have to make it generic unnecessarily.

```c#
   public class SomeService {
         public ISomeStruct ProcessInput(IContext context, ISomeStruct input) {
         }
   }
```

```c#

    // every class 

    public class SomeService<T>
          where T: struct, ISomeStruct {
         public T ProcessInput(IContext context, T input) {
               // do something with T ... 
         }
    }
```

Another larger problem is reflection itself, I use reflection to convert/serialize types, again I have to inspect generic type, extract parameter and then convert and pass it on. I have to unnecessarily box object with null even when not needed, because I do not know whether the target type is nullable or not. And someone sends a null value when it shouldn't be.

Custom analyzer is again a huge pain !! It is unnecessary code for simple solution. Provision is already  there in CLR. Purpose of compiler and language feature is to write smaller code for simple operations.

**Comment by 333fred** [comment]
Either you have to make it generic, or this won't work. IValueType cannot specify the size of the struct, as it has no idea, so either your code must be generic (in which case the struct constraint is your friend), or the struct must be boxed. There's no way around this.

**Comment by theunrepentantgeek** [comment]
@ackava - you should check out #164 - [*Exploration: Shapes and Extensions*](https://github.com/dotnet/csharplang/issues/164) for some of the ideas that the C# team have been considering.

As discussed in the above link, the implementation of shapes would allow you to write generic methods across multiple types (including structs) via the injection of _witness types_ that adapt the actual type to fit the required shape. Nicely, in many cases the witness types will optimize away completely at runtime, leaving optimal code.

**Comment by canton7** [comment]
However, my understanding is that you'd still need to use generics with shapes.

Essentially, either you use generics, which tell the runtime that the generic type parameter is variable-size and that it should emit specialised code for each type (if it's a value type); or you box the value type so that you get a reference of known size, and use that.

**Comment by ackava** [comment]
> Either you have to make it generic, or this won't work. IValueType cannot specify the size of the struct, as it has no idea, so either your code must be generic (in which case the struct constraint is your friend), or the struct must be boxed. There's no way around this.

This exactly is the reason IValueType which will have some hidden implementation of size in interface, like first byte will represent number of bytes of the value type.

**Comment by HaloFour** [comment]
There'd only be one `IValueType` interface, it can't have some "hidden implementation" which contains details of other arbitrary types.  And value types aren't limited to 255 bytes.  As mentioned, `IValueType` can't specify the size of the struct.  Generics are the required and appropriate solution here.  That is why the CLR specializes these generic members.

**Comment by CyrusNajmabadi** [comment]
Let me put the question this way:

What does this provide me that the shapes specification does not *already* provide?  It feels like a strict subset of that functionality.

**Comment by ackava** [comment]
@CyrusNajmabadi Problem is, generics does not work with interface which are implemented on platform specific code where app only accesses those features via interface. Consumer only has access to interface. Then in this case, value type will be boxed and boxing will add overhead of `GC`.

Take example of Xamarin, many things are only available as interface and they are implemented differently on different platform, even though we know, we cannot use valuetype as everything will always pass as reference, even for simplest thing such as passing RGBA color value will require boxing. On mobile, there are two virtual environment, first native VM (swift/obj-c) and JVM in Android and alone with there is .NET runtime, most of time some value just moves around stack and gets destroyed.

There could also be a fixed size string, for most strings, we only compare and take a decision, we don't need to store them.

@HaloFour IL code does not care about implementation, IValueType can have a hidden size of structure and IL can copy and move bytes. I see no reason why it is not possible.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Problem is, generics does not work with interface which are implemented on platform specific code where app only accesses those features via interface

Why not?  You can be generic and constrain on an interface. The impl of the value types that implement that interface can live in platform specific modules

**Comment by CyrusNajmabadi** [comment]
>  I see no reason why it is not possible.

You might want to consider providing a proof of concept pr to demonstrate what you mean and show how this could be done.

**Comment by CyrusNajmabadi** [comment]
I don't actually understand how you want your proposal to work. What would happen with:

```
public ISomeStruct ProcessInput(IContext context, ISomeStruct input) {
         }
```

What would the imple of that method be?  How would this work given you could pass in arbitrary instances of arbitrary types into it?

**Comment by ackava** [comment]
@theunrepentantgeek I have seen shapes, but again, look at my updated description, consumer of the API does not have access to concrete implementations, for example when you call `List<int>` , you have `int` available to you, but what if you are in middleware and you have no idea of what actual type is and you still have to write all your logic.

**Comment by canton7** [comment]
I don't follow your edit. You say things like "1 Byte length if less than 127 or 2 byte length" -- where is that data actually stored? Remember that interfaces are reference types -- they are a word-sized pointer. 

Imagine you have a class `C` which has a field of type `IValueType`. How much space should every instance of `C` reserve for each `IValueType` field? With a traditional interface, there would be 4/8 bytes available, which is just enough for a pointer. You can't fit 65535 bytes of storage in a pointer. The only other alternative is to special-case `IValueType` so it's some magic structure which is always 65535+4+2 bytes large, but then every instance of `C` needs this massive inline 65541 bytes of storage space for each `IValueType` field, which is huge. That's not an attractive option.

**Comment by ackava** [comment]
@canton7 Yes I agree, I don't expect struct to be of that great size either, couple of bytes may be max.

**Comment by canton7** [comment]
@ackava I think you're unfortunately missing the point I'm trying to make.

The *only* option that I see for a field of type `IValueType` is for that field to be 65541 bytes in size. There is no other possible way of doing this.

When an object is JIT-compiled, the offsets of each of the fields, and therefore the sizes of each of the fields, needs to be known at that point. If you have a field of type `IValueType`, its size is *not* known at JIT-time.

Imagine you have:

```cs
public class C
{
    public IValueType Field;
}
```

The JIT has to decide how much memory to allocate for `Field`. Remember that the user could do `c.Field = new LargeStruct()` or `c.Field = new SmallStruct()`, or even both of these after each other. `Field` therefore has to have enough storage for the largest possible struct that would be stored in it, which is 65541 bytes, looking at your proposal.

Do you see the problem?

**Comment by ackava** [comment]
@canton7 I got your point, well in that case we can have `IValueBytes8` , `IValueBytes16`, `IValueBytes64`, fixed sized struct, where in JIT knows how much bytes to allocate + 4 byte of Type pointer. I have updated my answer.

**Comment by CyrusNajmabadi** [comment]
> I have updated my answer.

I don't see how this would work.  You mention that generics are not ok, because it would require that  "every method, every class I need to write will have unnecessary extra <T>."

But now, with your approach here (of `IValueBytes8`, `IValueBytes16`, `IValueBytes32`, `IValueBytes64`), you'll need overloads for every one of those versions so that you can pass that data along to everything. 

You will still need all the overloads.  Except that instead of just saying `T` and having the runtime stamp out the right jitted method for every case, you will manually be stamping out all the different versions and have to maintain them all yourself.

**Comment by ackava** [comment]
@CyrusNajmabadi First of all I will never use `IValueByte8` directly, I am only going to use my interfaces such as...

```c#
   public interface IColor: IValueBytes8 {
   }

    // ios
   public struct NSColor: IColor {
   }

   // droid
   public struct DroidColor: IColor {
   }

```

I will only be using `IColor` everywhere in my code, why would I need any different overload? JIT knows that `IColor` is `IValueBytes8` so it needs 8 + 4 bytes.

**Comment by CyrusNajmabadi** [comment]
If you know the size... why not just pass a single struct along?  it will be smaller and simpler.  i.e. you would only need to pass along however large your color struct actually is.

**Comment by CyrusNajmabadi** [comment]
> JIT knows that IColor is IValueBytes8 so it needs 8 + 4 bytes.

But how does this actually work?  so you call this with an NSColor in one case, and a DroidColor in another.  What does the runtime do in those cases as these structs have different layouts and functionality and whatnot.

**Comment by CyrusNajmabadi** [comment]
I also don't understand this bit:

> Problem is, CommonConsumer have to write too much of generics in order to maintain struct being converted to object. This becomes even more complicated that everything has to be generic even though not required.

Why do you have to wrte too much generics.  You'd ust need to write:  `T : where T : struct, IColor`.  And now you get an impl that has no overhead and is specialized properly for every possible impl you want to throw at it (regardless of size).  

Any code you need to write "in oder to maintain struct being converted to object" would still have to be written with yoru IValueTypeXX approach as well.

**Comment by CyrusNajmabadi** [comment]
Afaict, your proposal seems to be asking for a less powerful, and less efficient system than what can be done with generics today. 

The ways that it is less efficient:

- You have overhead in passing values around.  You have to:
  - pass a type-descriptor along with the value.
  - you need to pad your value up to the next IValueXXX size available.
  - all your calls need to go through an indirection of fetching the accessor through the type-descriptor and then invoking it on the data itself.

The ways that it is less powerful:

- You are limited only to the sizes specified by IValueBytesXXX.  Android colors, for example, can be 64bits, so you'd need at least IValueBytes64.  And that doesn't cover the fact that you could easily have cases that go beyond that.
- You no longer get method jitting that specializes for the actual struct being passed in.  This means you need extra space for all processing, and you don't get any sort of efficient codegen for each potential struct.
- You would not be able to have things like arrays work at all.  if i have an `IColor[]` that would necessarily mean that each entry in the array would have to have the space for the type-descriptor and value.  Which would not be compatible with something like an `NSColor[]`.  Contrast this with generics, where this just works (optimally).

You're reinventing a poor-mans generics.  One that would work for almost no purposes and would strictly works far less well than the system that already exists :-/

You eliminate this as an option in your OP, but i cannot actually understand the reasoning behind your rejection.  You say "See the amount of code changed from simple class to one with Generics", but then show code that doesn't seem like it was hard to change at all.  So i'm not getting what the problem is there.

**Comment by ackava** [comment]
@CyrusNajmabadi 

`IColor[]` would work exactly same way how `int[]` works. I don't see any difference. Array of `IColor[]` would be a reference to actual bytes values held by structs. I never asked `IColor[]` to be value type, it is object.

I would suggest, you try one application in `Xamarin.Forms` and try creating an API where you will eventually force user to use `T` as a concrete type and write everything in generics, see what happens, we can't write it, we avoid it and use objects, we only use reference types when we create cross platform libraries. Show me which API in Xamarin.Forms uses generics to enforce performance? "NONE" because consumer of API never references `NSColor` or `DroidColor`, how would it be platform independent code?

**Comment by canton7** [comment]
@ackava In X.F, can't you just use [`Xamarin.Forms.Color`](https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.color?view=xamarin-forms)? There are extension methods to turn this from and into platform-specific colors, e.g. `ColorExtensions.ToColor()` and `ColorExtensions.ToUIColor()`. No need for `IColor`.

Having done some largeish X.F projects, the slowness is hardly ever due to allocating a few Color-like objects. Run it through a profiler, and you'll probably find that the cost comes from your views: the extra objects required to map X.F view hierarchies into native view hierarchies add up quickly!

**Comment by ackava** [comment]
> @ackava In X.F, can't you just use [`Xamarin.Forms.Color`](https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.color?view=xamarin-forms)? There are extension methods to turn this from and into platform-specific colors, e.g. `ColorExtensions.ToColor()` and `ColorExtensions.ToUIColor()`. No need for `IColor`.
> 
> Having done some largeish X.F projects, the slowness is hardly ever due to allocating a few Color-like objects. Run it through a profiler, and you'll probably find that the cost comes from your views: the extra objects required to map X.F view hierarchies into native view hierarchies add up quickly!

I just gave `Xamarin.Forms.Color` as an example, of course I know it will not make much of performance difference, think of various other native integrations, V8, Graphics Libraries, .. Sqlite, ..

**Comment by canton7** [comment]
@ackava The general point still stands though. For things like Sqlite, you'll want to be dealing with classes not structs, so this issue doesn't apply. If you can dream up another scenario where you have some platform-dependent struct that you want to hide behind an interface, you can still probably create a platform-agnostic struct (like `X.F.Color`), and map to/from it as appropriate. This is going to be cheaper than your proposal, and more flexible (you can create an instance of a platform-agnostic struct in platform-agnostic code), and take a similar amount of boilerplate.

Do you have an actual, concrete example of a type which would benefit from your proposal? Include in your example some code which indicates that using generics is too painful to consider.

**Comment by CyrusNajmabadi** [comment]
> IColor[] would work exactly same way how int[] works. I don't see any difference. Array of IColor[] would be a reference to actual bytes values held by structs.

How would that work?  You said that IColors would be structs of a particular size. And you said that a type-descriptor would go with each value so you actually know what sort of instance you were working with. 

So your IColor instances are always a different (i.e. larger) size than the actual original struct value. 

So you cannot convert arrays like this. They are literally different sizes

**Comment by canton7** [comment]
I think the word "compatible" got lost in there, somewhere.

@ackava I think @CyrusNajmabadi is saying that, with generics, you can do:

```cs
void Foo<T>(T[] arr) where T : IColor { }
...
NSColor[] arr = ...
Foo(arr);
```

But you wouldn't be able to do the same with your proposal. You can't have a method which takes an `IColor[]`, and pass it an `NSColor[]`. You'd need to construct a new array, and copy each `NSColor` into it.

**Comment by CyrusNajmabadi** [comment]
Yup.  That's exactly it @canton7 .  And why i was mentioning that this approach would be both less efficient and less powerful than what we already have today.

**Comment by 333fred** [comment]
Another thing that these specialized types would do is make introducing a new field to a struct a source-breaking change, as the struct would no longer satisfy whatever size container you are shoving it into. I do not see any advantages to this over `where T : struct`, only limitations and disadvantages.

**Comment by svick** [comment]
I wonder if a different approach would make sense here: conditional compilation (either on the file level or using `#if`s) based on OS (identified using an [RID](https://docs.microsoft.com/en-us/dotnet/core/rid-catalog)). This way, you don't need the abstraction of interfaces, because you have a separate .dll for each OS.

CoreFX is already written this way and NuGet supports such packages, so I think the main hurdle would be to set up the project system to make this work in your projects. That might not be trivial, but it could be okay for performance critical code.
